Add optional query param for adding headers to the server's response

moritzschmitz-oviva opened the pull request:
Add a feature to allow setting response headers via query params.

Derived from #31.

moritzschmitz-oviva committed:
Add optional query param for adding headers to the server's response

moritzschmitz-oviva committed:
Add optional query param for adding headers to the server's response

moritzschmitz-oviva committed:
Formatting

moritzschmitz-oviva committed:
Formatting

moritzschmitz-oviva committed:
Formatting

moritzschmitz-oviva committed:
Formatting

moritzschmitz-oviva committed:
Use header instead

moritzschmitz-oviva committed:
Use header instead

ldez commented on app.go:
```diff
@@ -219,6 +219,18 @@ func whoamiHandler(w http.ResponseWriter, r *http.Request) {
 
+	if headers, exists := queryParams["header"]; exists {
+		for _, header := range headers {
+			keyValue := strings.Split(header, "=")
+			if len(keyValue) != 2 {
+				continue
+			}
+			headerName := keyValue[0]
+			headerValue := keyValue[1]
+			w.Header().Set(headerName, headerValue)
```
```suggestion
			headerName, headerValue, found := strings.Cut(header, "=")
			if !found {
				continue
			}
			
			w.Header().Set(headerName, headerValue)
```

moritzschmitz-oviva replied:
Nice one. Thank you!

ldez commented:
Note, this doesn't fix #31 because this issue was not related to query parameters but to a request without any arguments.
The issue was about CLI flags.

moritzschmitz-oviva committed:
Add test

moritzschmitz-oviva committed:
Add test

moritzschmitz-oviva committed:
Applied suggestion and added a test for it

moritzschmitz-oviva committed:
Applied suggestion and added a test for it

ldez commented on app_test.go:
```diff
@@ -0,0 +1,63 @@
+			[]string{"Content-Type", "Cache-Control"},
+			[]string{"application%2Fjson", "no-store"},
+			[]string{"application/json", "no-store"},
+		},
+		{"duplicate headers",
+			[]string{"Cache-Control", "Cache-Control"},
+			[]string{"no-cache", "no-store"},
+			[]string{"no-store", "no-store"},
+		},
+		{"special encoded headers",
+			[]string{"X-Custom-Field"},
+			[]string{"equal%3Done"},
+			[]string{"equal=one"},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Given
+			params := make([]string, len(test.names))
+			for idx, part := range test.names {
+				params = append(params, fmt.Sprintf("header=%s=%s", part, test.setValues[idx]))
+			}
+			url := fmt.Sprintf("/random-path?%s", strings.Join(params, "&"))
+
+			// When
+			req := httptest.NewRequest(http.MethodGet, url, nil)
+			w := httptest.NewRecorder()
+			setHeaderFilter(w, req)
+
+			// Then
+			for idx, part := range test.names {
+				responseHeader := w.Header().Get(part)
+				if responseHeader != test.expectedValues[idx] {
+					t.Errorf("%s: expected %s got %s", part, test.expectedValues[idx], responseHeader)
+				}
+			}
+		})
+	}
+}
```
```suggestion
func TestApp_setHeaderFilter(t *testing.T) {
	tests := []struct {
		desc           string
		params         map[string][]string
		expectedValues []string
		expected       http.Header
	}{
		{
			desc:           "single header",
			params:         map[string][]string{"Content-Type": {"application/json"}},
			expectedValues: []string{"application/json"},
			expected:       http.Header{"Content-Type": []string{"application/json"}},
		},
		{
			desc: "multiple headers",
			params: map[string][]string{
				"Content-Type":  {"application/json"},
				"Cache-Control": {"no-store"},
			},
			expectedValues: []string{"application/json", "no-store"},
			expected: http.Header{
				"Content-Type":  []string{"application/json"},
				"Cache-Control": []string{"no-store"},
			},
		},
		{
			desc: "duplicate headers",
			params: map[string][]string{
				"Content-Type":  {"application/json"},
				"Cache-Control": {"no-store", "no-cache"},
			},
			expected: http.Header{
				"Content-Type":  []string{"application/json"},
				"Cache-Control": []string{"no-store", "no-cache"},
			},
		},
		{
			desc: "special encoded headers",

			params: map[string][]string{
				"X-Custom-Field": {"equal=one"},
			},
			expected: http.Header{
				"X-Custom-Field": []string{"equal=one"},
			},
		},
	}

	for _, test := range tests {
		t.Run(test.desc, func(t *testing.T) {
			t.Parallel()

			uri, err := url.Parse("/random-path")
			if err != nil {
				t.Fatal(err)
			}

			query := uri.Query()

			for k, values := range test.params {
				for _, v := range values {
					query.Add("header", fmt.Sprintf("%s=%s", k, v))
				}
			}

			uri.RawQuery = query.Encode()

			req := httptest.NewRequest(http.MethodGet, uri.String(), nil)
			w := httptest.NewRecorder()
			setHeaderFilter(w, req)

			for k, v := range w.Header() {
				if slices.Compare(test.expected.Values(k), v) != 0 {
					t.Errorf("header %s doesn't match: got %v, expected %v", k, v, test.expected.Values(k))
				}
			}
		})
	}
}
```

ldez commented on app.go:
```diff
@@ -208,6 +208,20 @@ func dataHandler(w http.ResponseWriter, r *http.Request) {
+			if !found {
+				continue
+			}
+			w.Header().Set(headerName, headerValue)
```
```suggestion

			w.Header().Add(headerName, headerValue)
```

moritzschmitz-oviva committed:
Apply suggestions from code review

Co-authored-by: Ludovic Fernandez <ldez@users.noreply.github.com>

moritzschmitz-oviva committed:
Apply suggestions from code review

Co-authored-by: Ludovic Fernandez <ldez@users.noreply.github.com>

moritzschmitz-oviva committed:
Fix README.md, plus formatting

moritzschmitz-oviva committed:
Fix README.md, plus formatting

moritzschmitz-oviva commented:
@ldez Thank you for the input!

moritzschmitz-oviva committed:
Naming

moritzschmitz-oviva committed:
Naming

moritzschmitz-oviva committed:
Remove superfluous fields

moritzschmitz-oviva committed:
Remove superfluous fields

moritzschmitz-oviva committed:
Rename test

moritzschmitz-oviva committed:
Rename test

moritzschmitz-oviva committed:
Merge branch 'feature/add-response-header-param' of https://github.com/moritzschmitz-oviva/whoami into feature/add-response-header-param

moritzschmitz-oviva committed:
Fix test according to linter

moritzschmitz-oviva committed:
Merge branch 'master' into feature/add-response-header-param

moritzschmitz-oviva commented:
@mmatur I think your review got stale?

moritzschmitz-oviva commented:
@ldez @mmatur Anything I can do for this PR?

moritzschmitz-oviva committed:
feat: add optional query param for adding headers to response

Add `?header=key=value` query parameter support to the root handler.
Multiple headers can be added by repeating the param, e.g.
`/?header=Cache-Control=no-cache&header=X-Foo=bar`.
